File: GovCidadao/app/services/audit_service.py
"""Trilha de auditoria corporativa — grava em apicemiterio.auditlogs (unificado com API Node)."""
from __future__ import annotations

import asyncio
import logging
import re
from datetime import datetime, timezone
from typing import Any

from bson import ObjectId
from fastapi import Request
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorCollection

from app.core.config import settings
from app.deps import CurrentUser

logger = logging.getLogger(__name__)

AUDIT_MODULE = "gov_cidadao"

SENSITIVE_KEYS = frozenset(
    {
        "password",
        "confirmpassword",
        "confirm_password",
        "token",
        "jwt",
        "authorization",
        "refreshtoken",
        "refresh_token",
    }
)
PII_MASK_KEYS = frozenset({"cpf", "cpf_cnpj", "email", "phone", "reporter_contact"})

EVENT_TYPES = frozenset(
    {
        "CREATE",
        "UPDATE",
        "DELETE",
        "VIEW",
        "DOWNLOAD",
        "UPLOAD",
        "LOGIN",
        "LOGOUT",
        "APPROVE",
        "REJECT",
        "PERMISSION_CHANGE",
        "SECURITY",
        "OTHER",
    }
)

_audit_client: AsyncIOMotorClient | None = None
_audit_collection: AsyncIOMotorCollection | None = None


def mask_value(key: str, value: Any) -> Any:
    if value is None:
        return value
    k = key.lower()
    if k in SENSITIVE_KEYS:
        return "[redacted]"
    if k in PII_MASK_KEYS:
        s = str(value)
        if k == "email" and "@" in s:
            local, domain = s.split("@", 1)
            return f"{local[:2]}***@{domain}"
        if len(s) <= 4:
            return "***"
        return f"***{s[-4:]}"
    if isinstance(value, str) and len(value) > 500:
        return f"{value[:500]}…"
    return value


def sanitize_metadata(data: dict[str, Any] | None) -> dict[str, Any]:
    if not data:
        return {}
    out: dict[str, Any] = {}
    for k, v in data.items():
        if v is None:
            continue
        if k.lower() in SENSITIVE_KEYS:
            continue
        out[k] = mask_value(k, v)
    return out


def build_changes(
    before: dict[str, Any] | None,
    after: dict[str, Any] | None,
    fields: list[str] | None = None,
) -> list[dict[str, Any]]:
    before = before or {}
    after = after or {}
    keys = fields or sorted(set(before) | set(after))
    changes: list[dict[str, Any]] = []
    for field in keys:
        b = before.get(field)
        a = after.get(field)
        if b == a:
            continue
        changes.append(
            {
                "campo": field,
                "antes": mask_value(field, b),
                "depois": mask_value(field, a),
            }
        )
    return changes


def infer_event_type(action: str, explicit: str | None = None) -> str:
    if explicit and explicit in EVENT_TYPES:
        return explicit
    a = action.lower()
    if "login" in a:
        return "LOGIN"
    if "logout" in a:
        return "LOGOUT"
    if ".create" in a or a.endswith("_create") or "register" in a:
        return "CREATE"
    if ".delete" in a or "erase" in a:
        return "DELETE"
    if ".update" in a or "toggle" in a:
        return "UPDATE"
    if ".read" in a or ".list" in a or ".view" in a or "export" in a:
        return "VIEW"
    if "denied" in a or "security" in a or "invalid" in a or "failed" in a:
        return "SECURITY"
    return "OTHER"


def parse_client(request: Request | None) -> dict[str, Any]:
    if request is None:
        return {}
    h = request.headers
    return {
        k: v
        for k, v in {
            "app": (h.get("x-client-app") or h.get("x-app-id") or "").strip() or None,
            "platform": (h.get("x-client-platform") or "").strip() or None,
            "version": (h.get("x-client-version") or "").strip() or None,
            "screen": (h.get("x-screen-id") or h.get("x-client-screen") or "").strip() or None,
            "moduleHint": (h.get("x-client-module") or "").strip() or None,
            "requestId": (h.get("x-request-id") or "").strip() or None,
        }.items()
        if v is not None
    }


def _parse_user_agent(ua: str) -> dict[str, str]:
    s = ua or ""
    browser = (
        "Edge"
        if "Edg/" in s
        else "Chrome"
        if "Chrome/" in s
        else "Firefox"
        if "Firefox/" in s
        else "Safari"
        if "Safari/" in s and "Chrome" not in s
        else "Unknown"
    )
    os_name = (
        "Android"
        if "Android" in s
        else "iOS"
        if "iPhone" in s or "iPad" in s
        else "Windows"
        if "Windows" in s
        else "macOS"
        if "Mac OS X" in s or "Macintosh" in s
        else "Linux"
        if "Linux" in s
        else "Unknown"
    )
    return {"browser": browser, "os": os_name, "raw": s[:512]}


def _client_ip(request: Request | None) -> str | None:
    if request is None:
        return None
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        return forwarded.split(",")[0].strip()
    if request.client:
        return request.client.host
    return None


def _maybe_object_id(value: str | None) -> ObjectId | None:
    if not value or not re.fullmatch(r"[a-fA-F0-9]{24}", value):
        return None
    try:
        return ObjectId(value)
    except Exception:
        return None


def _build_actor(actor: CurrentUser | dict[str, Any] | None) -> dict[str, Any]:
    if actor is None:
        return {}
    if isinstance(actor, CurrentUser):
        data = {
            "id": actor.id,
            "name": actor.name,
            "email": actor.email,
            "role": actor.role.value if hasattr(actor.role, "value") else str(actor.role),
        }
    else:
        data = dict(actor)
    oid = _maybe_object_id(str(data.get("id") or data.get("_id") or ""))
    out: dict[str, Any] = {
        "actorName": data.get("name"),
        "actorRole": data.get("role"),
        "actorEmail": data.get("email"),
    }
    if oid is not None:
        out["actorId"] = oid
    return out


async def init_audit_db() -> None:
    global _audit_client, _audit_collection
    uri = (settings.external_mongo_uri or settings.mongo_uri or "").strip()
    if not uri:
        logger.warning("[audit] URI Mongo ausente — trilha Gov desabilitada")
        return
    _audit_client = AsyncIOMotorClient(uri)
    _audit_collection = _audit_client[settings.audit_db][settings.audit_collection]
    logger.info("[audit] GovCidadao → %s.%s", settings.audit_db, settings.audit_collection)


async def close_audit_db() -> None:
    global _audit_client, _audit_collection
    if _audit_client:
        _audit_client.close()
    _audit_client = None
    _audit_collection = None


async def _write_log(
    request: Request | None,
    *,
    action: str,
    resource_type: str,
    resource_id: str | None = None,
    status: str = "success",
    metadata: dict[str, Any] | None = None,
    event_type: str | None = None,
    actor: CurrentUser | dict[str, Any] | None = None,
    changes: list[dict[str, Any]] | None = None,
    module: str = AUDIT_MODULE,
) -> None:
    if _audit_collection is None:
        return
    try:
        client = parse_client(request)
        ua = _parse_user_agent(request.headers.get("user-agent", "") if request else "")
        ip = _client_ip(request)
        now = datetime.now(timezone.utc)
        doc: dict[str, Any] = {
            **_build_actor(actor),
            "action": action,
            "resourceType": resource_type,
            "status": status,
            "metadata": sanitize_metadata(metadata),
            "ip": ip,
            "userAgent": request.headers.get("user-agent") if request else None,
            "module": module,
            "eventType": infer_event_type(action, event_type),
            "tenant": settings.audit_tenant,
            "client": client or None,
            "geo": {"ip": ip, "userAgent": ua.get("raw"), "browser": ua.get("browser"), "os": ua.get("os")},
            "route": str(request.url.path) if request else None,
            "method": request.method if request else None,
            "requestId": client.get("requestId"),
            "createdAt": now,
            "updatedAt": now,
        }
        if resource_id:
            doc["resourceId"] = str(resource_id)
        if changes:
            doc["changes"] = changes
        await _audit_collection.insert_one(doc)
    except Exception as exc:  # noqa: BLE001
        logger.error("[audit] Falha ao registrar trilha Gov: %s", exc)


def schedule_audit(coro) -> None:
    try:
        asyncio.get_running_loop().create_task(coro)
    except RuntimeError:
        pass


async def record_audit(
    request: Request | None,
    *,
    action: str,
    resource_type: str,
    resource_id: str | None = None,
    metadata: dict[str, Any] | None = None,
    event_type: str | None = None,
    actor: CurrentUser | dict[str, Any] | None = None,
    module: str = AUDIT_MODULE,
) -> None:
    await _write_log(
        request,
        action=action,
        resource_type=resource_type,
        resource_id=resource_id,
        metadata=metadata,
        event_type=event_type,
        actor=actor,
        module=module,
    )


async def record_security(
    request: Request | None,
    *,
    action: str,
    resource_type: str = "session",
    resource_id: str | None = None,
    metadata: dict[str, Any] | None = None,
    actor: CurrentUser | dict[str, Any] | None = None,
    status: str = "denied",
) -> None:
    await _write_log(
        request,
        action=action,
        resource_type=resource_type,
        resource_id=resource_id,
        metadata=metadata,
        event_type="SECURITY",
        actor=actor,
        status=status,
    )


async def record_change(
    request: Request | None,
    *,
    action: str,
    resource_type: str,
    resource_id: str | None = None,
    before: dict[str, Any] | None = None,
    after: dict[str, Any] | None = None,
    fields: list[str] | None = None,
    metadata: dict[str, Any] | None = None,
    event_type: str | None = None,
    actor: CurrentUser | dict[str, Any] | None = None,
) -> None:
    changes = build_changes(before, after, fields)
    await _write_log(
        request,
        action=action,
        resource_type=resource_type,
        resource_id=resource_id,
        metadata=metadata,
        event_type=event_type or "UPDATE",
        actor=actor,
        changes=changes or None,
    )


def audit_fire(coro) -> None:
    """Fire-and-forget — não bloqueia a resposta HTTP."""
    schedule_audit(coro)

File: backend/helpers/audit-service.js
const crypto = require('crypto')
const AuditLog = require('../models/AuditLog')

const SENSITIVE_KEYS = new Set([
  'password',
  'confirmpassword',
  'token',
  'jwt',
  'authorization',
  'refreshtoken',
  'refreshToken',
  'emailverifytoken',
  'resetpasswordtoken',
])

const PII_MASK_KEYS = new Set(['cpf', 'cpf_cnpj', 'email', 'phone'])

const EVENT_TYPES = new Set([
  'CREATE',
  'UPDATE',
  'DELETE',
  'VIEW',
  'DOWNLOAD',
  'UPLOAD',
  'LOGIN',
  'LOGOUT',
  'APPROVE',
  'REJECT',
  'PERMISSION_CHANGE',
  'SECURITY',
  'OTHER',
])

function maskValue(key, value) {
  if (value === undefined || value === null) return value
  const k = String(key).toLowerCase()
  if (SENSITIVE_KEYS.has(k)) return '[redacted]'
  if (PII_MASK_KEYS.has(k)) {
    const s = String(value)
    if (k === 'email' && s.includes('@')) {
      const [local, domain] = s.split('@')
      return `${local.slice(0, 2)}***@${domain}`
    }
    if (s.length <= 4) return '***'
    return `***${s.slice(-4)}`
  }
  if (typeof value === 'string' && value.length > 500) {
    return `${value.slice(0, 500)}…`
  }
  return value
}

function sanitizeMetadata(input = {}) {
  const out = {}
  for (const [k, v] of Object.entries(input)) {
    if (v === undefined) continue
    if (SENSITIVE_KEYS.has(String(k).toLowerCase())) continue
    out[k] = maskValue(k, v)
  }
  return out
}

function parseUserAgent(ua = '') {
  const s = String(ua)
  const browser = /Edg\//.test(s)
    ? 'Edge'
    : /Chrome\//.test(s)
      ? 'Chrome'
      : /Firefox\//.test(s)
        ? 'Firefox'
        : /Safari\//.test(s) && !/Chrome/.test(s)
          ? 'Safari'
          : 'Unknown'
  const os = /Android/.test(s)
    ? 'Android'
    : /iPhone|iPad/.test(s)
      ? 'iOS'
      : /Windows/.test(s)
        ? 'Windows'
        : /Mac OS X|Macintosh/.test(s)
          ? 'macOS'
          : /Linux/.test(s)
            ? 'Linux'
            : 'Unknown'
  return { browser, os, raw: s.slice(0, 512) }
}

function parseClient(req) {
  const h = req?.headers || {}
  const app = (h['x-client-app'] || h['x-app-id'] || '').trim() || undefined
  const platform = (h['x-client-platform'] || '').trim() || undefined
  const version = (h['x-client-version'] || '').trim() || undefined
  const screen = (h['x-screen-id'] || h['x-client-screen'] || '').trim() || undefined
  const moduleHint = (h['x-client-module'] || '').trim() || undefined
  const requestId = (h['x-request-id'] || '').trim() || undefined
  return {
    app,
    platform,
    version,
    screen,
    moduleHint,
    requestId,
  }
}

function inferModule({ module, resourceType, path = '', client }) {
  if (module) return module
  if (client?.moduleHint) return client.moduleHint
  const p = String(path).toLowerCase()
  if (p.includes('/pets') || p.includes('/adoption-requests')) return 'garca_pet'
  if (p.includes('/sepultados') || p.includes('/dloc')) return 'memorial'
  if (p.includes('/arvores') || p.includes('/denounces') || p.includes('/vacinacao') || p.includes('/castracao')) {
    return 'sama'
  }
  if (p.includes('/votacao')) return 'votacao'
  if (p.includes('/forms-garca')) return 'forms'
  if (p.includes('/lgpd')) return 'lgpd'
  if (p.includes('/garca-cidadao') || p.includes('/occurrences')) return 'gov_cidadao'
  if (p.includes('/users') || p.includes('/auth')) return 'auth'
  const rt = String(resourceType || '').toLowerCase()
  if (rt.startsWith('pet') || rt === 'adoption_request') return 'garca_pet'
  if (rt === 'sepultado' || rt === 'user') return rt === 'user' ? 'auth' : 'memorial'
  if (rt === 'votation' || rt === 'vote' || rt.startsWith('voting_')) return 'votacao'
  if (rt.startsWith('tree') || rt === 'denounce' || rt === 'vaccination') return 'sama'
  return 'api'
}

function inferEventType(action = '', explicit) {
  if (explicit && EVENT_TYPES.has(explicit)) return explicit
  const a = String(action).toLowerCase()
  if (a.includes('login')) return 'LOGIN'
  if (a.includes('logout')) return 'LOGOUT'
  if (a.includes('.create') || a.endsWith('_create') || a.includes('register')) return 'CREATE'
  if (a.includes('.delete') || a.includes('erase')) return 'DELETE'
  if (a.includes('.update') || a.includes('toggle') || a.includes('assign')) return 'UPDATE'
  if (a.includes('.read') || a.includes('.list') || a.includes('get_')) return 'VIEW'
  if (a.includes('denied') || a.includes('security') || a.includes('invalid')) return 'SECURITY'
  if (a.includes('upload')) return 'UPLOAD'
  if (a.includes('forgot') || a.includes('reset')) return 'UPDATE'
  return 'OTHER'
}

function buildActor(req, override = {}) {
  const actor = { ...(req?.user || {}), ...override }
  return {
    actorId: actor._id || actor.id || undefined,
    actorName: actor.name || undefined,
    actorRole: actor.role || undefined,
    actorEmail: actor.email || undefined,
  }
}

function buildChanges(before = {}, after = {}, fields = null) {
  const keys = fields || [...new Set([...Object.keys(before), ...Object.keys(after)])]
  const changes = []
  for (const field of keys) {
    const b = before[field]
    const a = after[field]
    if (JSON.stringify(b) === JSON.stringify(a)) continue
    changes.push({
      campo: field,
      antes: maskValue(field, b),
      depois: maskValue(field, a),
    })
  }
  return changes
}

function filesFromMulter(files) {
  if (!files) return []
  const list = Array.isArray(files) ? files : [files]
  return list
    .filter(Boolean)
    .map((f) => ({
      name: f.originalname || f.filename,
      path: f.filename || f.path,
      type: f.mimetype,
      size: f.size,
      hash: f.buffer
        ? crypto.createHash('sha256').update(f.buffer).digest('hex')
        : undefined,
    }))
}

async function writeLog(req, payload = {}) {
  try {
    const client = parseClient(req)
    const path = req?.originalUrl || req?.path || ''
    const ua = parseUserAgent(req?.headers?.['user-agent'])
    const actor = buildActor(req, payload.actor || {})
    const moduleName = inferModule({
      module: payload.module,
      resourceType: payload.resourceType,
      path,
      client,
    })
    const eventType = inferEventType(payload.action, payload.eventType)
    const changes = payload.changes || (payload.before && payload.after
      ? buildChanges(payload.before, payload.after, payload.fields)
      : undefined)

    await AuditLog.create({
      ...actor,
      action: payload.action || 'unknown_action',
      resourceType: payload.resourceType || 'unknown_resource',
      resourceId: payload.resourceId ? String(payload.resourceId) : undefined,
      status: payload.status || 'success',
      metadata: sanitizeMetadata(payload.metadata || {}),
      ip: req?.ip,
      userAgent: req?.headers?.['user-agent'],
      module: moduleName,
      eventType,
      tenant: payload.tenant || process.env.AUDIT_TENANT || 'prefeitura-garca',
      changes: changes?.length ? changes : undefined,
      files: payload.files?.length ? payload.files : undefined,
      sessionId: payload.sessionId,
      requestId: client.requestId,
      client,
      geo: { ip: req?.ip, userAgent: ua.raw, browser: ua.browser, os: ua.os },
      route: path,
      method: req?.method,
    })
  } catch (err) {
    console.error('[audit] Falha ao registrar trilha:', err?.message || err)
  }
}

async function recordAudit(req, payload = {}) {
  return writeLog(req, payload)
}

async function recordSecurity(req, payload = {}) {
  return writeLog(req, {
    ...payload,
    status: payload.status || 'denied',
    eventType: payload.eventType || 'SECURITY',
  })
}

async function recordChange(req, payload = {}) {
  const changes = buildChanges(payload.before, payload.after, payload.fields)
  return writeLog(req, {
    ...payload,
    changes,
    eventType: payload.eventType || 'UPDATE',
  })
}

module.exports = {
  recordAudit,
  recordSecurity,
  recordChange,
  buildChanges,
  maskValue,
  sanitizeMetadata,
  inferModule,
  inferEventType,
  filesFromMulter,
  parseClient,
}
